src/cressi_goa.c

```diff
@@ -377,6 +381,47 @@ cressi_goa_device_set_fingerprint (dc_device_t *abstract, const unsigned char da
 	return DC_STATUS_SUCCESS;
 }
 
+struct cressi_goa_version_differences {
+	unsigned int version;
+	unsigned int idlen;
+	unsigned int logbook_entry_len;
+	unsigned int logbook_fp_offset;
+	unsigned int dive_fp_offset;
+	unsigned char cmd_logbook;
+};
+
+static dc_status_t
+cressi_goa_determine_version(dc_device_t *abstract, unsigned char *id, const struct cressi_goa_version_differences **conf)
+{
+	static const struct cressi_goa_version_differences version_differences[] = {
+		{ 0,  9, SZ_HEADER, FP_OFFSET, FP_OFFSET, CMD_LOGBOOK },
+		/*    4 is the new version
+		 *   11 is the new response length to the `CMD_VERSION` command
+		 *   15 is the length of an entry in the Logbook
+		 *    3 is the offset to the START date in the Logbook header
+		 *    9 is the offset to the START date in the Dive header
+		 * 0x23 is the new command to request the Logbook
+		 */
+		{ 4, 11,        15,         3,         9, CMD_LOGBOOK_V4 },
+	};
+	cressi_goa_device_t *device = (cressi_goa_device_t *) abstract;
+	dc_status_t status = DC_STATUS_SUCCESS;
+	unsigned char id_[11] = {0};
+	if (id == NULL)
+		id = id_;
+	for (size_t version = 0; version < C_ARRAY_SIZE(version_differences); version++) {
+		status = cressi_goa_device_transfer (device, CMD_VERSION, NULL, 0, id, version_differences[version].idlen, NULL, NULL);
+		if (status == DC_STATUS_SUCCESS) {
+			*conf = &version_differences[version];
+			return status;
+		}
+	}
+	if (status != DC_STATUS_SUCCESS) {
+		ERROR (abstract->context, "Failed to read the version information.");
+	}
+	return status;
+}
```

Comment:
I'm not sure if this runtime decision on the version is really necessary or if it would make more sense to have separate entry points for the two different versions.

The idea behind this was that some Goa-compatible DC maybe gets a firmware update and now talks v4 of the protocol. With this solution that would still be fine. But maybe it's also YAGNI

One thing is for sure: with this solution we create an Error message in the log on the first try on v4 DCs.

```
[15.561219] ERROR: Unexpected payload size (11). [in cressi_goa.c:168 (cressi_goa_device_receive)]
```

Reply:
In my opinion, reducing the number of correct choices that a user has to make for something to work will result in improved user friendliness, which is almost always desirable.
So, if there is no reason that a user might want to deliberately choose one or the other model irrespective of the hardware / firmware they have, and if it is possible to make the correct choice for the user based on information that we have, then I think the choice should be made for the user instead of expecting them to make the correct choice.

As for the error message, this is a bit ugly, and will create red herrings and confusion when users are trying to diagnose and report problems that they are having - so it should be avoided if possible.

Reply:
I also prefer auto-detection of the protocol variant. Those error messages are not really a problem. They are mainly intended for developers and not for end-users.

But I think there is a better solution possible. The `cressi_goa_device_transfer` function either downloads a fixed amount of data with `cressi_goa_device_receive` (CMD_VERSION) or a variable sized amount of data with `cressi_goa_device_download` (CMD_LOGBOOK and CMD_DIVE), but never both. That means we could remove the output and osize arguments and always return the data through the variable sized buffer instead. That way the `cressi_goa_device_receive` function can return the correct size, and thus we don't have to try with different sizes.

Note that this code will need some changes for the bluetooth support anyway.

Reply:
I don't understand why you think you need to preserve the output/osize parameters. The only case where the `cressi_goa_device_receive` function returns some data is for the CMD_VERSION command. For all the other commands, this functions returns a zero length response (and thus we set output=NULL, osize=0). The actual data will be received in the `cressi_goa_device_download` function (for the CMD_LOGBOOK and CMD_DIVE commands) or there is no response data at all (for the CMD_SET_TIME_V4 and CMD_EXIT_PCLINK).

Thus we can pass a dc_buffer_t to the `cressi_goa_device_receive` function. If it returns a zero-length answer, we can continue with the `cressi_goa_device_download` to read the response. If we don't expect any response at all, we can pass a NULL pointer and skip the `cressi_goa_device_download`. See commit 8aeedc64d88b81d15a60f28e56298979efbbed11 for the idea. If there would be another (future) command where this assumption is no longer valid, we can always pass two dc_buffer_t pointers. But I doubt that will happen because for the BLE communication the `cressi_goa_device_receive` function isn't even needed.

You are actually doing almost the same in your code, except in a much more complicated way. After your patch, the `cressi_goa_device_transfer` is never called with output != NULL (remember CMD_VERSION is the only command that needed it), so all the parameter checking and manipulation isn't necessary and makes the code only harder to understand.

Reply:
> The only case where the `cressi_goa_device_receive` function returns some data is for the CMD_VERSION command.

The `CMD_LOGBOOK` command in v4 returns the number of logbook entries in the response (c.f. [0] [1]), that's why I had to retain it.

> If there would be another [...] command where this assumption is no longer valid, we can always pass two dc_buffer_t pointers.

I can change it to have two `dc_buffer_t` pointers if you prefer that.

[0] https://sjaeckel.github.io/2024/04/23/Cressi-Nepto-for-Subsurface.html#cmd-0x23---logbook
[1] https://github.com/sjaeckel/libdivecomputer/blob/87a4ac5543dde87270bb148e11e540ba820593a9/src/cressi_goa.c#L536-L547

Reply:
Looks like I somehow completely missed that. I now wonder what that means for the BLE support, because it omits the entire `cressi_goa_device_receive` part. Maybe that's why the BLE communication uses yet another variant of the CMD_LOGBOOK and CMD_DIVE commands?

I think using two dc_buffer_t will result in simpler code. The silly thing is that we don't really need this number of logbook entries for anything. The total size is already communicated during the download.

Reply:
Done